File: 백준알고리즘/DFS(깊이 우선 탐색)/python/유기농 배추.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
cnt = 0


def solution():
    M, N, K = map(int, input().split())
    global cnt
    graph = [[0] * N for _ in range(M)]

    for _ in range(K):
        m1, m2 = map(int, input().split())
        graph[m1][m2] = 1

    def dfs(x, y):
        graph[x][y] = 0

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx < 0 or nx >= M or ny < 0 or ny >= N:
                continue
            if graph[nx][ny] == 1:
                dfs(nx, ny)

    for i in range(M):
        for j in range(N):
            if graph[i][j] == 1:
                cnt += 1
                dfs(i, j)

    return cnt


for _ in range(int(input())):
    cnt = 0
    print(solution())

File: 백준알고리즘/DFS(깊이 우선 탐색)/javascript/유기농 배추.js
const fs = require('fs');
const stdin = fs.readFileSync('/dev/stdin').toString().split('\n');
const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const T = +input();
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

for (let i = 0; i < T; i++) {
  const [M, N, K] = input()
    .split(' ')
    .map(n => +n);
  const graph = Array.from(Array(N), () => Array(M).fill(0));

  for (let i = 0; i < K; i++) {
    const [u, v] = input()
      .split(' ')
      .map(n => +n);
    graph[v][u] = 1;
  }

  function dfs(x, y) {
    graph[x][y] = 0;

    for (let i = 0; i < 4; i++) {
      const nx = x + dx[i];
      const ny = y + dy[i];

      if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;

      if (graph[nx][ny] === 1) dfs(nx, ny);
    }
  }

  let cnt = 0;
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (graph[i][j] === 1) {
        cnt++;
        dfs(i, j);
      }
    }
  }

  console.log(cnt);
}
